egs3/mini_an4/asr/dataset/builder.py

```
@@ -115,7 +118,11 @@ def prepare_source(self, recipe_dir: str | Path, **_kwargs) -> None:
         dataset_root = recipe_root / _CFG["dataset_path"]
         archive = (recipe_root / _CFG["archive_path"]).resolve()
         if not archive.exists():
-            raise FileNotFoundError(f"Archive not found: {archive}")
+            url = _CFG.get("archive_url", "")
+            if not url:
+                raise FileNotFoundError(f"Archive not found: {archive}")
+            logger.info("Downloading archive | url=%s -> %s", url, archive)
+            urllib.request.urlretrieve(url, str(archive))
```

Ruff (0.16.2)

[error] Audit URL open for permitted schemes. Allowing use of `file:` or custom schemes is often unexpected.

(S310)